### PR TITLE
PDAL TileDB plugin

### PR DESCRIPTION
**Pull request by normanb**

### CODE REVIEW

File: plugins/tiledb/io/TileDBReader.cpp [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+                if (nDims == 3)
+                    view->setField(Dimension::id("Z"), idx, (*spCoords)[i + 2]);
+
+                // read tiledb attributes, this has to be done point by point for future streaming support
+                for (const auto &kv : buffers) {
+                    if (kv.first != TILEDB_COORDS) {
+                        Dimension::Id id = Dimension::id(kv.first);
+                        Dimension::Type t = Dimension::defaultType(id);
+                        int p = i / 3;
```
Comment (tdenniston) [inline plugins/tiledb/io/TileDBReader.cpp]: Shouldn't this be `i / nDims`?

File: plugins/tiledb/io/TileDBReader.cpp [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+}
+
+point_count_t TileDBReader::read(PointViewPtr view, point_count_t count)
+{
+    PointId idx = view->size();
+    PointRef point(*view, idx);
+    point_count_t numRead = 0;
+    tiledb::Query query(*m_ctx, *m_array);
```
Comment (tdenniston) [inline plugins/tiledb/io/TileDBReader.cpp:113]: I didn't see anywhere here where a subarray is being set on the query -- should there be one set?

File: plugins/tiledb/io/TileDBReader.cpp (hunk 1 above)
Comment (normanb) [reply plugins/tiledb/io/TileDBReader.cpp]: +1, yes, I will make that change.

File: plugins/tiledb/io/TileDBReader.cpp (hunk 2 above)
Comment (normanb) [reply plugins/tiledb/io/TileDBReader.cpp:113]: Not in this case, the reader doesn't subset.

File: plugins/tiledb/io/TileDBReader.cpp [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+    if (key != TILEDB_COORDS)
+        query.set_buffer(key, *sp);
+}
+
+void TileDBReader::addArgs(ProgramArgs& args)
+{
+    args.add("array_name", "TileDB array name", m_arrayName).setPositional();
+    args.add("config_file", "TileDB configuration file location", m_cfgFileName);
+    args.add("chunk_size", "TileDB read chunk size", m_chunkSize, point_count_t(1000));
```
Comment (jakebolewski) [inline plugins/tiledb/io/TileDBReader.cpp]: Discussed extensions:
* Subarray support (optional)
* Stats output for benchmarking (optional)

File: plugins/tiledb/io/TileDBWriter.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    {
+        // The point view takes on the spatial reference of that stage, if it had one.
+        meta.add("spatialreference", pdal::Utils::toString(getSpatialReference()));
+    }
+
+    meta.add("bounds", pdal::Utils::toString(m_bbox));
+
+    // serialize pipeline
+    std::ostream* ostr = FileUtils::createFile(m_arrayName + ".json");
+    pdal::Utils::toJSON(anon, *ostr);
```
Comment (jakebolewski) [inline plugins/tiledb/io/TileDBWriter.cpp]: Discussed saving this with VFS to support all the backends of TileDB.

File: plugins/tiledb/io/TileDBWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        case Dimension::Type::Unsigned64:
+            setAttrAnyValue(mp, attrName, view->getFieldAs<unsigned long>(id, idx));
+            break;
+        case Dimension::Type::None:
+        default:
+            throw pdal_error("Unsupported attribute type for " + view->dimName(id));
+    }
+}
+
+tiledb::Attribute createAttribute(const tiledb::Context& ctx, const std::string name, Dimension::Type t)
```
Comment (tdenniston) [inline plugins/tiledb/io/TileDBWriter.cpp:120]: Should we use a default compressor for the attributes?

File: plugins/tiledb/io/TileDBWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        }
+    }
+
+    m_query->set_coordinates(coords);
+    m_query->submit();
+}
+
+void TileDBWriter::done(PointTableRef table)
+{
+    m_query->finalize();
```
Comment (tdenniston) [inline plugins/tiledb/io/TileDBWriter.cpp]: Finalize is only strictly needed for global order writes.

Comment (normanb) [reply plugins/tiledb/io/TileDBWriter.cpp]: :+1:

File: plugins/tiledb/io/TileDBWriter.cpp (hunk 4 above)
Comment (normanb) [reply plugins/tiledb/io/TileDBWriter.cpp]: :+1:

File: plugins/tiledb/io/TileDBReader.cpp (hunk 3 above)
Comment (normanb) [reply plugins/tiledb/io/TileDBReader.cpp]: :+1: added both

**Review by stavrospapadopoulos — APPROVED** [review]

File: plugins/tiledb/io/TileDBWriter.cpp (hunk 5 above)
Comment (normanb) [reply plugins/tiledb/io/TileDBWriter.cpp:120]: :+1: Added compression options to the writer

**Comment by normanb** [comment]
Before merge
- [ ] revert scripts/docker/master/ubuntu/Dockerfile to HEAD of master, currently changed so as to run a docker image of this branch
- [x] squash commits

Can add to scripts/docker/master/ubuntu/Dockerfile once this has been merged into PDAL master. Will revert docker changes in `pdal-tiledb-dev` branch

**Pull request merged**